### PR TITLE
Upgrade steps based on cluster state

### PR DESCRIPTION
**Pull request by sedroche**
Jira: https://issues.redhat.com/browse/OSD-3916

### CODE REVIEW

**Comment by sedroche** [comment]
I need to run a few more upgrades with this but initial runs were fine. Would appreciate any feedback

**Comment by sedroche** [comment]
For metrics to work this needs to be run as a deployment. An image is here with these changes -> quay.io/sedroche/managed-upgrade-operator:idempotentish

File: pkg/maintenance/alertmanagerMaintenance.go [hunk 1]
```diff
@@ -157,3 +186,14 @@ func createDefaultMatchers() []*amv2Models.Matcher {
 	inNamespaceAlertMatcher := createMatcher("namespace", "(^openshift.*|^kube.*|^redhat.*|^default$)", true)
 	return amv2Models.Matchers{nonCriticalAlertMatcher, inNamespaceAlertMatcher}
 }
+
+// Check if a silence with comment exists
+func silenceExists(mList *amSilence.GetSilencesOK, comment string) bool {
```
Comment (mrbarge) [inline pkg/maintenance/alertmanagerMaintenance.go:191]: If an expired silence exists will it still return true here? does it also need to check payload.Status? 
(haven't confirmed if this is an issue, just wondering)

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -67,57 +118,180 @@ func init() {
...
-func (c *Counter) UpdateMetricClusterVerificationSucceeded(upgradeconfig string) {
+func (c *Counter) UpdateMetricClusterVerificationSucceeded(upgradeConfigName string) {
 	metricClusterVerificationFailed.With(prometheus.Labels{
-		nameLabel: upgradeconfig}).Set(
-			float64(0))
+		nameLabel: upgradeConfigName}).Set(
+		float64(0))
+}
+
+func getPromUrl(c client.Client) (*string, error) {
```
Comment (mrbarge) [inline pkg/metrics/metrics.go:217]: Minor nit.. when I think 'url' I think of the form with the protocol and path included, eg 'https://blah'.. maybe rename to getPromHost() to be clearer about what it's returning?

**Comment by mrbarge** [comment]
I see where this is going and on the whole think it's a good change. I haven't tested it yet and want to do another pass over it now that I properly understand the goals, but so far it seems ok. Left some comments.

File: pkg/maintenance/alertmanagerMaintenance.go (hunk 1 above)
Comment (sedroche) [reply pkg/maintenance/alertmanagerMaintenance.go:191]: It does return expired silences. Each silence should only be created once and If it expires we get alerted for it.
If we check for active we can end up recreating a maintenance.

File: pkg/cluster_upgrader/cluster_upgrader.go [hunk 3]
```diff
@@ -279,34 +300,47 @@ func EnsureExtraUpgradeWorkers(c client.Client, metricsClient metrics.Metrics, m
...
+	if isEqualVersion(cv, upgradeConfig) {
+		logger.Info(fmt.Sprintf("ClusterVersion is already set to Channel %s Version %s, skipping %s", desired.Channel, desired.Version, upgradev1alpha1.CommenceUpgrade))
+		return true, nil
+	}
+
+	if cv.Spec.DesiredUpdate != nil &&
+		cv.Spec.DesiredUpdate.Version == upgradeConfig.Spec.Desired.Version &&
+		cv.Spec.Channel == upgradeConfig.Spec.Desired.Channel {
 		return true, nil
 	}
```
[the remark this answers is not in the document]
Comment (sedroche) [reply pkg/cluster_upgrader/cluster_upgrader.go:317]: Nice. Thanks.

File: pkg/cluster_upgrader/cluster_upgrader.go [hunk 4]
```diff
@@ -139,7 +140,17 @@ func Ordering() []upgradev1alpha1.UpgradeConditionType {
...
-	ok, err := performClusterHealthCheck(c, logger)
+	cv, err := getClusterVersion(c)
+	if err != nil {
+		return false, err
+	}
+	desired := upgradeConfig.Spec.Desired
+	if isEqualVersion(cv, upgradeConfig) {
+		logger.Info(fmt.Sprintf("ClusterVersion is already set to Channel %s Version %s, skipping %s", desired.Channel, desired.Version, upgradev1alpha1.UpgradePreHealthCheck))
+		return true, nil
+	}
```
[the remark this answers is not in the document]
Comment (sedroche) [reply pkg/cluster_upgrader/cluster_upgrader.go:151]: Sounds right. Not sure why I did it this way now.

File: pkg/metrics/metrics.go (hunk 2 above)
Comment (sedroche) [reply pkg/metrics/metrics.go:217]: :+1:

**Comment by sedroche** [comment]
@mrbarge Some updates have been pushed (can be seen in https://github.com/openshift/managed-upgrade-operator/compare/7fb0e9e73805587b5cfe1f3a4b89c2407203ca2d..37cd88ffdb33cfdb44bb1098faae17383e00b86c)

**Pull request merged**